lightly/models/utils.py

```diff
@@ -518,40 +521,15 @@ def nearest_neighbors(
 
     """
 
-    batch_size = input_maps.size(0)
-
     if num_matches is None or num_matches == -1:
         num_matches = input_maps.size(1)
 
-    topk_values, topk_indices = distances.topk(k=1, largest=False)
-    topk_values = topk_values.squeeze(-1)
-    topk_indices = topk_indices.squeeze(-1)
-
-    sorted_values, sorted_values_indices = torch.sort(topk_values, dim=1)
-    sorted_indices, sorted_indices_indices = torch.sort(sorted_values_indices, dim=1)
-
-    mask = torch.stack(
-        [
-            torch.where(sorted_indices_indices[i] < num_matches, True, False)
-            for i in range(batch_size)
-        ]
-    )
-    topk_indices_selected = topk_indices.masked_select(mask)
-
-    topk_indices_selected = topk_indices_selected.reshape(batch_size, num_matches)
-
-    indices = (
-        torch.arange(0, topk_values.size(1))
-        .unsqueeze(0)
-        .repeat(batch_size, 1)
-        .to(topk_values.device)
-    )
-    indices_selected = indices.masked_select(mask)
-    indices_selected = indices_selected.reshape(batch_size, num_matches)
-
-    filtered_input_maps = _batched_index_select(input_maps, 1, indices_selected)
-    filtered_candidate_maps = _batched_index_select(
-        candidate_maps, 1, topk_indices_selected
-    )
+    topk_values, topk_indices = distances.topk(k=1, dim=2, largest=False) # [bsz, map_size_0, 1]
+    topk_values = topk_values.squeeze(-1) # [bsz, map_size_0]
+    topk_indices = topk_indices.squeeze(-1) # [bsz, map_size_0]
+    min_values_1, min_indices_2 = topk_values.topk(k=num_matches, dim=1, largest=False) # [bsz, num_matches]
+    filtered_input_maps = _batched_index_select(input_maps, 1, min_indices_2) # [bsz, num_matches, num_input_maps]
```

Comment:
That's true, I guess I have to do a gather as for the candidate maps

Reply:
After some thought, I think it is correct. topk_values has the same shape and indexing of input_maps. It is just a max on distances in the maps lets say second dimension

Reply:
Hmm, maybe I don't understand something. Let's assume we have the following example:

```
topk_values = [0.1, 1.0, 0.2, 0.5]
topk_indices = [3, 1, 2, 0]
_, min_inidices = topk_values.topk(k=2)
```
Now we have `min_indices = [1, 3]`. But we cannot use this one the select the entries from the `input_maps` because `min_indices` is an index into `topk_values`. We know that that we want the entries with `topk_values=[1.0, 0.5]` which are at index `[1, 0]` in the `input_maps`. To get `[1, 0]` we have to do `topk_indices[min_indices]` no?

Reply:
Aah I see now! Thanks a lot for the explanation! Could you add maybe a couple more comments on what is happening?

Reply:
I'll try. But it is very difficult to visualize it and understand fully what's happening without a sketch 😂